apps/examples/src/examples/layer-panel/ShapeList.tsx

```diff
@@ -0,0 +1,149 @@
+import { capitalize } from 'lodash'
+import { useState } from 'react'
+import { Editor, TLShapeId, useEditor, useValue } from 'tldraw'
+import { VisibilityOff, VisibilityOn } from '../../icons/icons'
+
+const selectedBg = '#E8F4FE'
+const childSelectedBg = '#F3F9FE'
+const childBg = '#00000006'
+
+function ShapeItem({
+	shapeId,
+	depth,
+	parentIsSelected,
+	parentIsHidden,
+}: {
+	shapeId: TLShapeId
+	depth: number
+	parentIsSelected?: boolean
+	parentIsHidden?: boolean
+}) {
+	const editor = useEditor()
+
+	const shape = useValue('shape', () => editor.getShape(shapeId), [editor])
+	const children = useValue('children', () => editor.getSortedChildIdsForParent(shapeId), [editor])
+	const isHidden = useValue('isHidden', () => editor.isShapeHidden(shapeId), [editor])
+	const isSelected = useValue('isSelected', () => editor.getSelectedShapeIds().includes(shapeId), [
+		editor,
+	])
+	const shapeName = useValue('shapeName', () => getShapeName(editor, shapeId), [editor])
+
+	const [isEditingName, setIsEditingName] = useState(false)
+
+	if (!shape) return null
+
+	return (
+		<>
+			{!!shape && (
+				<div
+					className="shape-item"
+					onDoubleClick={() => {
+						setIsEditingName(true)
+					}}
+					onClick={() => {
+						// We synchronize the selection state of the layer panel items with the selection state of the shapes in the editor.
+						if (editor.inputs.ctrlKey || editor.inputs.shiftKey) {
+							if (isSelected) {
+								editor.deselect(shape)
+							} else {
+								editor.select(...editor.getSelectedShapes(), shape)
+							}
+						} else {
+							editor.select(shape)
+						}
+					}}
+					style={{
+						paddingLeft: 10 + depth * 20,
+						opacity: parentIsHidden || isHidden ? 0.5 : 1,
+						background: isSelected
+							? selectedBg
+							: parentIsSelected
+								? childSelectedBg
+								: depth > 0
+									? childBg
+									: undefined,
+					}}
+				>
+					{isEditingName ? (
+						<input
+							autoFocus
+							className="shape-name-input"
+							defaultValue={shapeName}
+							onBlur={() => setIsEditingName(false)}
+							onChange={(ev) => {
+								if (shape.type === 'frame') {
+									editor.updateShape({ ...shape, props: { ...shape.props, name: ev.target.value } })
```

Comment:
do we need to spread `shape.props` here? thought we could just pass in a partial of what we want and it would only update that, so to avoid race conditions spreading should be avoided, but you folks would know far better on that!

Reply:
yes you are correct! Supplying a partial is definitely preferred. I'll fix.